segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -86,59 +65,35 @@ public ReadIndexCache(int maxIndexedSegments, int maxIndexedChunksPerSegment, in
      * @param streamSegmentName Name of the segment.
      * @return Read index corresponding to the given segment. A new empty index is created if it doesn't already exist.
      */
-    private SegmentReadIndex getSegmentReadIndex(String streamSegmentName) {
-        SegmentReadIndex readIndex = segmentsToReadIndexMap.get(streamSegmentName);
-
-        if (null == readIndex) {
-            // Evict segments if required.
-            if (maxIndexedSegments < segmentsToReadIndexMap.size() + 1 || maxIndexedChunks < totalChunkCount.get() + 1) {
-                evictSegmentsFromOldestGeneration();
+    private SegmentReadIndex getSegmentReadIndex(String streamSegmentName, boolean createIfNotPresent) {
+        SegmentReadIndex readIndex = segmentsReadIndexCache.getIfPresent(streamSegmentName);
+        if (null == readIndex && createIfNotPresent) {
+            synchronized (segmentsReadIndexCache) {
```

Comment:
Guava cache is synchronized. Why are you re-synchronizing?

Reply:
Multiple threads might find that entry does not exist and might race to create the entry for the first time. (This code is invoked only once per segment)

I did not find clean putIfNotPresent equivalent method.

Reply:
SimpleCache has putIfAbsent. Check it out